src/metrics/metrics-tracer-factory.ts

```diff
@@ -0,0 +1,247 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as crypto from 'crypto';
+import * as os from 'os';
+import * as process from 'process';
+import {v4 as uuidv4} from 'uuid';
+import {Counter, Histogram, metrics} from '@opentelemetry/api';
+import {detectResourcesSync} from '@opentelemetry/resources';
+import {gcpDetector} from '@opentelemetry/resource-detector-gcp';
+import * as Constants from './constants';
+import {MetricsTracer} from './metrics-tracer';
+const version = require('../../../package.json').version;
+
+export class MetricsTracerFactory {
+  private static _instance: MetricsTracerFactory | null = null;
+  private _clientAttributes: {[key: string]: string};
+  private _instrumentAttemptCounter!: Counter;
+  private _instrumentAttemptLatency!: Histogram;
+  private _instrumentOperationCounter!: Counter;
+  private _instrumentOperationLatency!: Histogram;
+  private _instrumentGfeConnectivityErrorCount!: Counter;
+  private _instrumentGfeLatency!: Histogram;
+  public enabled: boolean;
+  public gfeEnabled: boolean;
+
+  private constructor(enabled: boolean, gfeEnabled = false) {
+    this.enabled = enabled;
+    this.gfeEnabled = gfeEnabled;
+    this._createMetricInstruments();
+
+    const clientName = `${Constants.SPANNER_METER_NAME}/${version}`;
+    const clientUid = MetricsTracerFactory._generateClientUId();
+    const clientHash = MetricsTracerFactory._generateClientHash(clientUid);
+    const location = MetricsTracerFactory._detectClientLocation();
+
+    this._clientAttributes = {
+      [Constants.METRIC_LABEL_KEY_CLIENT_NAME]: clientName,
+      [Constants.METRIC_LABEL_KEY_CLIENT_UID]: clientUid,
+      [Constants.MONITORED_RES_LABEL_KEY_CLIENT_HASH]: clientHash,
+      // Skipping instance config until we have a way to get it
+      [Constants.MONITORED_RES_LABEL_KEY_INSTANCE_CONFIG]: 'unknown',
+      [Constants.MONITORED_RES_LABEL_KEY_LOCATION]: location,
+    };
+  }
+
+  public static getInstance(enabled: boolean, gfeEnabled = false) {
```

Comment:
Should the service metrics and GFE metrics be enabled or disabled by default?  And what should the env vars be?

In Java I see:
```
SPANNER_DISABLE_BUILTIN_METRICS
SPANNER_DISABLE_DIRECT_ACCESS_GRPC_BUILTIN_METRICS
```

Which sounds like they should be enabled by default and the env vars can be set to turn them off.